### PR TITLE
feat(SPV-794): contacts api

### PR DESCRIPTION
**Pull request by mariusz-4chain**
# Pull Request Checklist

- [ ] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [ ] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [ ] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [ ] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [ ] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: actions/contacts/search.go [hunk 1]
```diff
@@ -72,3 +72,116 @@ func (a *Action) search(c *gin.Context) {
 
 	c.JSON(http.StatusOK, response)
 }
+
+// TODO: this method will be changed based on search poc
```
Comment (chris-4chain) [inline actions/contacts/search.go:76]: Are you planning to do it in this PR or in some next task?
- If in the next task, is there a JIRA ticket for it?

File: actions/contacts/search.go [hunk 2]
```diff
@@ -72,3 +72,116 @@ func (a *Action) search(c *gin.Context) {
...
+
+	reqParams.DefaultsIfNil()
+
+	contacts, err := a.Services.SpvWalletEngine.GetContactsByXpubID(
+		c.Request.Context(),
+		reqXPubID,
+		mappings.MapToMetadata(reqParams.Metadata),
+		conditions,
+		mappings.MapToQueryParams(reqParams.QueryParams),
+	)
```
Comment (chris-4chain) [inline actions/contacts/search.go:175]: By `ID` or by `paymail` 🤔 ?

File: actions/contacts/unconfirm.go [hunk 3]
```diff
@@ -41,3 +41,20 @@ func (a *Action) unconfirm(c *gin.Context) {
 	}
 	c.Status(http.StatusOK)
 }
+
+// unconfirmContact will unconfirm contact request
+// Unconfirm contact godoc
```
Comment (chris-4chain) [inline actions/contacts/unconfirm.go:46]: What does `"godoc"` mean in this context

File: actions/contacts/search.go (hunk 1 above)
Comment (mariusz-4chain) [reply actions/contacts/search.go:76]: In next PR. This is [ticket number](https://bitcoinassociation.atlassian.net/browse/SPV-881)

File: actions/contacts/search.go (hunk 2 above)
Comment (mariusz-4chain) [reply actions/contacts/search.go:175]: Thanks. Fixed. It should be by `paymail`

File: actions/contacts/unconfirm.go (hunk 3 above)
Comment (mariusz-4chain) [reply actions/contacts/unconfirm.go:46]: It's pretty useless in this context. Removed.
It should mark godoc comment, but You know that.

File: actions/contacts/confirm.go [hunk 4]
```diff
@@ -41,3 +41,20 @@ func (a *Action) confirm(c *gin.Context) {
 	}
 	c.Status(http.StatusOK)
 }
+
+// confirmContact will confirm contact request
+// Confirm contact godoc
```
Comment (chris-4chain) [inline actions/contacts/confirm.go:46]: Same `"godoc"` here.

File: actions/contacts/upsert.go [hunk 5]
```diff
@@ -59,3 +59,18 @@ func (a *Action) upsert(c *gin.Context) {
 
 	c.JSON(http.StatusOK, response)
 }
+
+// upsertContact will add a new contact or modify an existing one.
+// upsertContact contact godoc
```
Comment (chris-4chain) [inline actions/contacts/upsert.go:64]: Unncessary `"godoc"`

File: actions/contacts/confirm.go [hunk 6]
```diff
@@ -41,3 +41,20 @@ func (a *Action) confirm(c *gin.Context) {
...
+// @Summary		Confirm contact
+// @Description	Confirm contact. For contact with status "unconfirmed" change status to "confirmed"
+// @Tags		Contacts
+// @Produce		json
+// @Param		paymail path string true "Paymail address of the contact the user wants to confirm"
+// @Success		200
+// @Failure		404	"Contact not found"
+// @Failure		422	"Contact status not unconfirmed"
+// @Failure		500	"Internal server error"
+// @Router		/v1/contacts/{paymail}/confirmation [PATCH]
```
Comment (dorzepowski) [inline actions/contacts/confirm.go:56]: Wouldn't it look better if it would be a POST

File: actions/contacts/unconfirm.go [hunk 7]
```diff
@@ -41,3 +41,20 @@ func (a *Action) unconfirm(c *gin.Context) {
...
+// @Summary		Unconfirm contact
+// @Description	Unconfirm contact. For contact with status "confirmed" change status to "unconfirmed"
+// @Tags		Contacts
+// @Produce		json
+// @Param		paymail path string true "Paymail address of the contact the user wants to unconfirm"
+// @Success		200
+// @Failure		404	"Contact not found"
+// @Failure		422	"Contact status not confirmed"
+// @Failure		500	"Internal server error"
+// @Router		/v1/contacts/{paymail}/non-confirmation [PATCH]
```
Comment (dorzepowski) [inline actions/contacts/unconfirm.go:56]: Hmm... It doesn't look the best,
maybe, if we use POST for confirmation, then we could change this to:
`DELETE /api/v1/contacts/{paymail}/confirmation`
WDYT?

File: actions/contacts/unconfirm.go [hunk 8]
```diff
@@ -41,3 +41,20 @@ func (a *Action) unconfirm(c *gin.Context) {
 	}
 	c.Status(http.StatusOK)
 }
+
+// unconfirmContact will unconfirm contact request
+// Unconfirm contact godoc
+// @Summary		Unconfirm contact
```
Comment (dorzepowski) [inline actions/contacts/unconfirm.go:47]: Same as for accept: use single method adjust description and use // @DeprecatedRouter

File: actions/contacts/upsert.go [hunk 9]
```diff
@@ -59,3 +59,18 @@ func (a *Action) upsert(c *gin.Context) {
 
 	c.JSON(http.StatusOK, response)
 }
+
+// upsertContact will add a new contact or modify an existing one.
```
Comment (dorzepowski) [inline actions/contacts/upsert.go:63]: Same as for accept: use single method adjust description and use // @DeprecatedRouter

File: models/response_model.go [hunk 10]
```diff
@@ -0,0 +1,13 @@
+package models
+
+type PageDescription struct {
```
Comment (dorzepowski) [inline models/response_model.go:3]: You should add comments (go doc) on both structs and fields so swagger could prepare documentation of it.

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: actions/contacts/confirm.go (hunk 6 above)
Comment (mariusz-4chain) [reply actions/contacts/confirm.go:56]: Changed.

File: actions/contacts/confirm.go (hunk 4 above)
Comment (mariusz-4chain) [reply actions/contacts/confirm.go:46]: Fixed.

File: actions/contacts/confirm.go [hunk 11]
```diff
@@ -41,3 +41,20 @@ func (a *Action) confirm(c *gin.Context) {
 	}
 	c.Status(http.StatusOK)
 }
+
+// confirmContact will confirm contact request
```
[the remark this answers is not in the document]
Comment (mariusz-4chain) [reply actions/contacts/confirm.go:45]: I used the same trick as in transactions API. @Deprecated with descritpion. @DeprecatedRouter doesn't work as expected.

File: actions/contacts/upsert.go (hunk 5 above)
Comment (mariusz-4chain) [reply actions/contacts/upsert.go:65]: Fixed

File: actions/contacts/upsert.go (hunk 9 above)
Comment (mariusz-4chain) [reply actions/contacts/upsert.go:64]: I used @Deprecated with description about replacement.

Comment (mariusz-4chain) [reply actions/contacts/upsert.go:64]: I used @Deprecated with description inside @Summary

File: actions/contacts/accept.go [hunk 12]
```diff
@@ -34,3 +34,20 @@ func (a *Action) accept(c *gin.Context) {
...
 	c.Status(http.StatusOK)
 }
+
+// acceptInvitations will accept contact request
+// Accept contact invitation godoc
+// @Summary		Accept contact invitation
+// @Description	Accept contact invitation. For contact with status "awaiting" change status to "unconfirmed"
+// @Tags		Contacts
+// @Produce		json
+// @Param		paymail path string true "Paymail address of the contact the user wants to accept"
```
Comment (Nazarii-4chain) [inline actions/contacts/accept.go:44]:
```suggestion
// @Param		paymail path string true "Paymail address of the contact that the user would like to accept"
```

File: actions/contacts/confirm.go [hunk 13]
```diff
@@ -21,7 +20,24 @@ import (
...
 func (a *Action) confirm(c *gin.Context) {
+	a.confirmContact(c)
+}
+
+// confirmContact will confirm contact request
+// @Summary		Confirm contact
+// @Description	Confirm contact. For contact with status "unconfirmed" change status to "confirmed"
+// @Tags		Contacts
+// @Produce		json
+// @Param		paymail path string true "Paymail address of the contact the user wants to confirm"
```
Comment (Nazarii-4chain) [inline actions/contacts/confirm.go:33]:
```suggestion
// @Param		paymail path string true "Paymail address of the contact that the user would like to confirm"
```

File: actions/contacts/reject.go [hunk 14]
```diff
@@ -33,3 +33,20 @@ func (a *Action) reject(c *gin.Context) {
...
 	c.Status(http.StatusOK)
 }
+
+// rejectInvitation will reject contact request
+// Reject contact invitation godoc
+// @Summary		Reject contact invitation
+// @Description	Reject contact invitation. For contact with status "awaiting" delete contact
+// @Tags		Contacts
+// @Produce		json
+// @Param		paymail path string true "Paymail address of the contact the user wants to reject"
```
Comment (Nazarii-4chain) [inline actions/contacts/reject.go:43]:
```suggestion
// @Param		paymail path string true "Paymail address of the contact that the user would like to reject"
```

File: actions/contacts/upsert.go [hunk 15]
```diff
@@ -15,16 +15,31 @@ import (
...
 func (a *Action) upsert(c *gin.Context) {
+	a.upsertContact(c)
+}
+
+// upsertContact will add a new contact or modify an existing one.
+// @Summary		Upsert contact
+// @Description	Add or update contact. When adding a new contact, the system utilizes Paymail's PIKE capability to dispatch an invitation request, asking the counterparty to include the current user in their contacts.
+// @Tags		Contacts
+// @Produce		json
+// @Param		paymail path string true "Paymail address of the contact the user wants to add/modify"
```
Comment (Nazarii-4chain) [inline actions/contacts/upsert.go:37]:
```suggestion
// @Param		paymail path string true "Paymail address of the contact that the user would like to add/modify"
```

File: models/response_model.go [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+	Size          int `json:"size"`
+	Number        int `json:"number"`
+	TotalElements int `json:"totalElements"`
+	TotalPages    int `json:"totalPages"`
+}
+
+type PageModel[T any] struct {
+	Content []*T            `json:"content"`
```
Comment (Nazarii-4chain) [inline models/response_model.go:11]: Maybe `data` would be a better name. WDYT?

File: actions/contacts/accept.go (hunk 12 above)
Comment (yarex-4chain) [reply actions/contacts/accept.go]: changed

File: actions/contacts/confirm.go (hunk 13 above)
Comment (yarex-4chain) [reply actions/contacts/confirm.go]: changed

File: actions/contacts/reject.go (hunk 14 above)
Comment (yarex-4chain) [reply actions/contacts/reject.go]: changed

File: actions/contacts/upsert.go (hunk 15 above)
Comment (yarex-4chain) [reply actions/contacts/upsert.go]: changed

File: models/response_model.go (hunk 16 above)
Comment (yarex-4chain) [reply models/response_model.go]: I'm not sure - both `content` and `data` seem correct. `content` was used in the previous PRs

File: models/response_model.go (hunk 10 above)
Comment (yarex-4chain) [reply models/response_model.go]: this file was removed, but I will add the comments on the `response/response_model.go`

File: actions/contacts/upsert.go (hunk 9 above)
Comment (yarex-4chain) [reply actions/contacts/upsert.go]: done (single method adjust description and use // @DeprecatedRouter)

File: actions/contacts/accept.go [hunk 17]
```diff
@@ -42,3 +42,20 @@ func (a *Action) accept(c *gin.Context) {
...
+// @Success		200
+// @Failure		404	"Contact not found"
+// @Failure		422	"Contact status not awaiting"
+// @Failure		500	"Internal server error"
+// @Router		/v1/invitations/{paymail} [POST]
+// @Security	x-auth-xpub
```
[the remark this answers is not in the document]
Comment (yarex-4chain) [reply actions/contacts/accept.go:40]: done

File: actions/contacts/confirm.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (yarex-4chain) [reply actions/contacts/confirm.go:28]: done (single method adjust description and use // @DeprecatedRouter)

File: actions/contacts/reject.go [hunk 18]
```diff
@@ -41,3 +41,20 @@ func (a *Action) reject(c *gin.Context) {
...
+// @Success		200
+// @Failure		404	"Contact not found"
+// @Failure		422	"Contact status not awaiting"
+// @Failure		500	"Internal server error"
+// @Router		/v1/invitations/{paymail} [DELETE]
+// @Security	x-auth-xpub
```
[the remark this answers is not in the document]
Comment (yarex-4chain) [reply actions/contacts/reject.go:40]: done (single method adjust description and use // @DeprecatedRouter)

File: actions/contacts/unconfirm.go (hunk 8 above)
Comment (yarex-4chain) [reply actions/contacts/unconfirm.go:30]: done (single method adjust description and use // @DeprecatedRouter)

File: actions/contacts/routes.go [hunk 19]
```diff
@@ -30,3 +30,27 @@ func NewHandler(appConfig *config.AppConfig, services *config.AppServices) route
...
+		group := router.Group("/contacts")
+		group.PUT("/:paymail", action.upsertContact)
+		group.PATCH("/:paymail/confirmation", action.confirmContact)
+		group.PATCH("/:paymail/non-confirmation", action.unconfirmContact)
+
+		group.GET("", action.getContacts)
+		group.GET(":paymail", action.getContactsByID)
+	})
+
+	invitationsAPIEndpoints := routes.APIEndpointsFunc(func(router *gin.RouterGroup) {
```
[the remark this answers is not in the document]
Comment (yarex-4chain) [reply actions/contacts/routes.go:48]: It was done this way in other PRs

File: actions/contacts/routes.go [hunk 20]
```diff
@@ -30,3 +30,27 @@ func NewHandler(appConfig *config.AppConfig, services *config.AppServices) route
 
 	return apiEndpoints
 }
+
+// NewContactsHandler creates the specific package routes
+func NewContactsHandler(appConfig *config.AppConfig, services *config.AppServices) (routes.APIEndpointsFunc, routes.APIEndpointsFunc) {
+	action := &Action{actions.Action{AppConfig: appConfig, Services: services}}
+
+	contactsAPIEndpoints := routes.APIEndpointsFunc(func(router *gin.RouterGroup) {
```
[the remark this answers is not in the document]
Comment (yarex-4chain) [reply actions/contacts/routes.go:38]: done

[the remark this answers is not in the document]
Comment (yarex-4chain) [reply actions/contacts/routes.go:38]: done

File: actions/contacts/search.go [hunk 21]
```diff
@@ -72,3 +72,116 @@ func (a *Action) search(c *gin.Context) {
...
+// @Summary		Get contacts by id
+// @Description	Get contacts by id
+// @Tags		Contacts
+// @Produce		json
+// @Success		200 {object} models.Contact "Contact"
+// @Failure		400	"Bad request - Error while parsing SearchContacts from request body"
+// @Failure 	500	"Internal server error - Error while searching for contacts"
+// @Router		/v1/contacts/{id} [GET]
+// @Security	x-auth-xpub
+func (a *Action) getContactsByID(c *gin.Context) {
```
[the remark this answers is not in the document]
Comment (yarex-4chain) [reply actions/contacts/search.go]: done

**Review by wregulski — APPROVED** [review]

File: models/response_model.go (hunk 10 above)
Comment (yarex-4chain) [reply models/response_model.go]: added comments

File: actions/contacts/unconfirm.go (hunk 7 above)
Comment (yarex-4chain) [reply actions/contacts/unconfirm.go]: after some thought - what you proposed is better (POST "creates" a confirmation and DELETE "removes" it and `non-confirmation` seems ambigious). 
So I changed it

File: actions/contacts/routes.go [hunk 22]
```diff
@@ -12,21 +12,45 @@ type Action struct {
...
+		group.DELETE("/:paymail/confirmation", action.unconfirmContact)
+
+		group.GET("", action.getContacts)
+		group.GET(":paymail", action.getContactByPaymail)
+	})
+
+	invitationsAPIEndpoints := routes.APIEndpointsFunc(func(router *gin.RouterGroup) {
+		group := router.Group("/invitations")
+
+		group.POST("/:paymail", action.acceptInvitations)
```
Comment (dorzepowski) [inline actions/contacts/routes.go:51]: according to the JIRA task this should be:
```suggestion
		group.POST("/:paymail/contacts", action.acceptInvitations)
```

File: actions/contacts/routes.go [hunk 23]
```diff
@@ -12,21 +12,45 @@ type Action struct {
...
 
 	return apiEndpoints
 }
+
+// NewHandler creates the specific package routes
+func NewHandler(appConfig *config.AppConfig, services *config.AppServices) (routes.APIEndpointsFunc, routes.APIEndpointsFunc) {
+	action := &Action{actions.Action{AppConfig: appConfig, Services: services}}
+
+	contactsAPIEndpoints := routes.APIEndpointsFunc(func(router *gin.RouterGroup) {
+		group := router.Group("/contacts")
```
Comment (dorzepowski) [inline actions/contacts/routes.go:39]: There is no DELETE contact endpoint, which seem to be reasonable operation (and is listed in JIRA task)

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: actions/contacts/routes.go (hunk 22 above)
Comment (yarex-4chain) [reply actions/contacts/routes.go]: changed

File: actions/contacts/routes.go (hunk 23 above)
Comment (yarex-4chain) [reply actions/contacts/routes.go:39]: I've created an implementation of the DELETE contact endpoint

Comment (yarex-4chain) [reply actions/contacts/routes.go:39]: but I'm not sure if you can find a contact ID from the paymail in a simpler way, that does not require passing an empty JSON or skipping the `Content-Type: application/json` header for the endpoint

Comment (yarex-4chain) [reply actions/contacts/routes.go:39]: I've changed it - it required changes in the engine, but now works as it should

**Review by dorzepowski — APPROVED** [review]

**Pull request merged**